### DISCUSSION TITLE
Best practice for apps that have long execution times?

### DISCUSSION THREAD

**Discussion by jwberck** [nasa/cFS#510] [Q&A]
For some cases, there is a need to allow an app to run for multiple seconds processing some dense task. Typically this is some sort of GN&C application. When using SCH wakeups in CFS, there are a couple issues that come up with this sort of problem. If you run the app at a frequency that is faster than it's processing time, wakeup messages build up, resulting in strange unwanted behavior. It seems there are a couple different strategies I've seen for handling these cases, I'd like to know if anyone else has run into this and what you ended up doing. Are there any best practices to draw from?

1. Define upper bound on expected execution time, set SCH wakeup period to be above this value
    - This is the simplest solution, and seems to work well for applications that occur periodically
    - The major con here is this app is needed, there is potential latency in the system that is equal to the size of the period defined in Sch
    - That con is the most apparent in apps that are not called very frequently, where they may be needed and will require seconds to pass before getting another wakeup call
2. Instead of using SCH wakeup, key off of some other input
    - This addresses the con above by being called immediately by some caller app when the application is needed. This reduces the latency to nothing, but you don't get the periodic wakeups from SCH that sort of define your high level behavior. 
    -  This seems to be a better solution for apps that are not called frequently and are not periodically defined. However, for apps that are periodic, the strategy laid out in point 1 seems like a better fit
3. Track app frequency/period within the app, clear out the wakeup pipes in the app being woken up if it gets wakeups before it's period expires
    -  This seems like the dirtiest solution, as it pulls a lot of knowledge about scheduling into the app itself and require it to handle throwing out wakeup calls. However, it does provide protection against cases where multiple wakeups may be sent anomalously. This could only happen if that app is blowing it's minor frame or if it's init function is still not finished before SCH starts sending wakeups. It guarantees multiple anomalous wakeups won't accidentally be acted on and doesn't require defining maximum run times and such, which could be potentially useful for smaller, lesser funded projects.
    
   
That's the ideas I've seen so far to tackle this problem, I'd love to hear someone else's perspective and if there are best practices defined here.

**Comment by ghost** [comment]
Most of our cFS based missions use a periodic housekeeping request at ~5 seconds. Some apps have 1 second wake up messages as well.
Some apps have long background tasks run in a cFS Child Task (thread) where it does not need to respond to software bus messages. You can see examples of this in the File Manager (FM) app.

I cannot claim to be a GNC expert, but I thought most of our GNC systems run at greater than 1Hz, which means they would always be available to process data from a software bus pipe.

I would think that if every cycle was multiple seconds, then it might make sense to see if the algorithm could be broken up to allow the app to remain responsive to the software bus scheduling and housekeeping requests.
If the processing that takes multiple seconds is an asynchronous activity, like the FM example, then offloading it to a child task may be an approach that would work.

As an example, our Stored Command app can execute an absolute time sequence and multiple relative time sequences simultaneously , but the sequences are not implemented as threads. It's more like a single threaded state machine where it's driven by a 1hz scheduling message from the software bus. It follows the pattern: wake up, process messages, do work, go back to wait for messages.

Alan

**Reply by jwberck** [reply]
Ah yes I forgot about the child task pattern. We considered using this for an application that is not periodic and runs for multiple seconds, but ended up keying off another app for its "wakeup" while still doing housekeeping at a longer rate. The primary question we ran into that left us unsatisfied with that particular solution was "If the child task is doing basically everything except consuming a wakeup and hk command, what benefit are we really gaining from this pattern"? It seemed like, we could get hk back during processing at a faster rate, but the hk to report wound 't be of much importance since there wouldn't be anything new to report until the child thread finished, so it would really only allow command counts and such to be updated. Knowing if the app is stuck or alive could be achieved with HS, so we decided to move away from that pattern in this case. I'm curious what benefits you found from that pattern, and will have to take a look at FM, I didn't realize it used that pattern. I'd expect if the child thread has more reporting power, or if there are multiple, then that could be more useful.

Breaking up the app processing into multiple stages is definitely a good idea, that's something I think could be applied in a lot of cases. However, that are some cases that make that quite difficult. The main one that I'm thinking of is splitting up a deep neural network for a computer vision system. It could be done, but would probably result in a less efficient implementation and require some jumping through hoops as these are typically implemented in python and then converted to a defined C or C++ format via tools like tflite. 

Good suggestion with SC too as an example, I'll take a look. For that case, would the assumption be that as long as all the relative time sequences are <1s and 1 sequence is executed each SC frame?

Thanks for the response, that's good food for thought, I'll dig into those apps a bit more.

**Reply by leomeza1** [reply]
I recently started researching how to implement these GN&C functions in cFS and found this discussion. Have you had any breakthroughs or good answers? Some GN&C functions need to run at high frequency (20Hz). I was thinking maybe there's a way to implement this stuff in the OSAL or even the PSP layer, but I'm new to cFS and haven't found a good solution yet.

**Reply by jwberck** [reply]
Depends on your usecase. If they are actually running at 20Hz you can just use the normal scheduler. If you need 20Hz for some sort of reporting on the progress of a longer running algorithm (say on the order of seconds), then you'll want to use one of the described approaches like the child thread.

Personally, we did not find the child thread approach worth pursuing for our own case and we just have the app key off of an input that is not the scheduler wakeup to have it start as soon as possible (it's not a frequently recurring app, more of a one off).

I'm not entirely sure what you would use OSAL or PSP for with regard to a 20Hz GN&C app, if you describe your context more we might be able to provide better advice.

**Comment by ghost** [comment]
That is a good point about not having anything to report while the child task is busy. The only benefit I could see is having the app remain responsive - if it is possible to report progress or check for cancellation requests, the app could respond at the same rate as other apps. For example knowing that an operation is in progress and it is xx% finished may be useful in real time telemetry.

But maybe we should have a mindset of apps that run on a much longer cycle. Having "black box" algorithms or other subsystems in linux may also provide a use case for this, as it might not be practical to modify the code.
Alan

**Reply by jwberck** [reply]
Right, I think the issue there is it's only the top thread that is truly responsive unless your child can do some message passing between them. The main operation seems like the ability to kill the child thread, which we can get through a HS or AFM type app that is keeping track. I think there's something to be said for containing that kill operation within the app though, it's probably easier to recover from internally. It comes down to you FDIR scope and recovery strategy at that point I think.

Defining the treatment of these longer cycle black box algorithms would definitely be good for the community I think, that's really what I'm hoping to understand better.
